Python/Product/Analysis/LanguageServer/Settings.cs

```diff
@@ -15,24 +15,49 @@
 // permissions and limitations under the License.
 
 using System;
+using System.Collections.Generic;
 
 namespace Microsoft.PythonTools.Analysis.LanguageServer {
-    internal class LanguageServerSettings {
-        public AutoCompleteSettings autoComplete;
-        public DiagnosticsSettings diagnostics;
+    public sealed class LanguageServerSettings {
+        public class PythonAnalysisOptions {
+            public bool openFilesOnly;
+            public string[] errors = Array.Empty<string>();
+            public string[] warnings = Array.Empty<string>();
+            public string[] information = Array.Empty<string>();
+            public string[] disabled = Array.Empty<string>();
 
-        [Serializable]
-        internal class AutoCompleteSettings {
-            public bool showAdvancedMembers;
-        }
+            private Dictionary<string, DiagnosticSeverity> _map;
+
+            public DiagnosticSeverity GetEffectiveSeverity(string code, DiagnosticSeverity defaultSeverity) {
+                Init();
```

Comment:
Pretty sure this can race (e.g. multiple parses completing on separate threads).

Reply:
It can race, but I don't see any issue with this dictionary filled multiple times. We can use `ConcurrentDictionary` or simply assign the instance to the field at the end of the method to avoid concurrent access before dictionary is filled. The `EditorFile._lock` is widely used object, but it is still scoped to one file, while `PythonAnalysisOptions` is shared for the whole server.

Reply:
We're going to end up filling the dictionary in probably 99.9% of instances, so we may as well do it as soon as we have the new values.

The problem is with a partially filled dictionary, which is either going to randomly NRE, or show diagnostics to the user that they explicitly disabled.

Reply:
Setting value to the `PythonAnalysisOptions._map` field after dictionary is fully filled solves the problem with partially filled dictionary.

Reply:
Initializing it once in the constructor solves that and the multiple-creation problem introduced by doing it lazily without a lock.

Reply:
The map should not be changing inside the loop or user may get some codes translated to old settings and some to new.

Reply:
Added explicit initialization. It shouldn't need lock since settings are replaced wholesale and then are only read.